core/embed/upymod/modtrezorcrypto/modtrezorcrypto-tropic.h

```
@@ -0,0 +1,172 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#if USE_TROPIC
+
+#include <sec/secret.h>
+#include <sec/tropic_transport.h>
+#include "libtropic.h"
+
+extern STATIC lt_handle_t lt_handle;
+
+/// package: trezorcrypto.tropic
+
+/// class TropicError(Exception):
+///     """Error returned by the Tropic Square chip."""
+MP_DEFINE_EXCEPTION(TropicError, Exception)
+
+#define PING_MSG_MAX_LEN 64
+#define ECC_SLOT_COUNT 32
+#define SIG_SIZE 64
+
+/// def ping(message: str) -> str:
+///     """
+///     Test the session by pinging the chip.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_tropic_ping(mp_obj_t message) {
+  lt_ret_t ret = LT_FAIL;
+
+  uint8_t msg_in[PING_MSG_MAX_LEN] = {0};
+
+  mp_buffer_info_t message_b = {0};
+  mp_get_buffer_raise(message, &message_b, MP_BUFFER_READ);
+  if (message_b.len >= 0) {
+    ret = lt_ping(&lt_handle, (uint8_t *)message_b.buf, (uint8_t *)msg_in,
+                  message_b.len);
+    if (ret != LT_OK) {
+      mp_raise_msg(&mp_type_TropicError, "lt_ping failed.");
+    }
+  } else {
+    return mp_const_none;
```

I would rather raise an exception here.